### PR TITLE
[WIP] Added HA details to fix #2454

### LINKED ISSUE #2454: Create docs on high-availability components and leader election
**Describe the change you'd like to see**
- Describe how HA works for Knative
- Provide details about which components are HA
- Default settings, how to configure HA

**Additional context**
- Based on https://docs.openshift.com/container-platform/4.3/serverless/serverless-config-HA.html
- Based on previous PR https://github.com/knative/docs/pull/2279 by @pmorie

### PR DESCRIPTION
**Pull request by abrennan89**
Fixes #2454

## Proposed Changes

- Added details about HA and leader election
- List of currently HA components
- How to disable HA

### CODE REVIEW

File: docs/install/config-ha.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+type: "docs"
+---
+
+Active/passive high availability (HA) is a standard feature of Kubernetes APIs that helps to ensure that APIs stay operational if a disruption occurs. In an HA deployment, if an active controller crashes or is deleted, another controller is available to take over processing of the APIs that were being serviced by the controller that is now unavailable.
+
+Active/passive HA in Knative is available through leader election, which is enabled by default after the Knative Serving control plane is installed.
+
+When using a leader election HA pattern, instances of controllers are already scheduled and running inside the cluster before they are required. These controller instances compete to use a shared resource, known as the leader election lock. The instance of the controller that has access to the leader election lock resource at any given time is referred to as the leader.
+
+HA functionality is available by default on Knative for the `autoscaler-hpa`, `controller`, `activator` , `kourier-control`, and `kourier-gateway` components. These components are configured with two replicas by default.
```
Comment (vagababov) [inline docs/install/config-ha.md:14]: I think `kourier` is redundant here? Or we need to list istio as well (both pilot and IGW are HA components).

File: docs/install/config-ha.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+<!-- Q: is this correct? What way do we want to format names like Activator? Does this depend on context?-->
+
+## Disabling high availability
+
+You can disable HA for Knative Serving components by changing the value of `KnativeServing.spec.highAvailability` from `2` (default) to `1`.
```
Comment (vagababov) [inline docs/install/config-ha.md:19]: Is this operator config?

File: docs/install/config-ha.md (hunk 1 above)
Comment (abrennan89) [reply docs/install/config-ha.md:14]: We could list Istio too. Is the component just `istio-gateway`? @vagababov

Comment (vagababov) [reply docs/install/config-ha.md:14]: both istio-gateway and istio-pilot are H(P)A components

Comment (vagababov) [reply docs/install/config-ha.md:14]: `HA functionality is available by default on Knative for the ` this is also a bit hard to parse:
since it might be percieved as _default on_ which it isn't.

File: docs/install/config-ha.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+type: "docs"
+---
+
+Active/passive high availability (HA) is a standard feature of Kubernetes APIs that helps to ensure that APIs stay operational if a disruption occurs. In an HA deployment, if an active controller crashes or is deleted, another controller is available to take over processing of the APIs that were being serviced by the controller that is now unavailable.
+
+Active/passive HA in Knative is available through leader election, which is enabled by default after the Knative Serving control plane is installed.
+
+When using a leader election HA pattern, instances of controllers are already scheduled and running inside the cluster before they are required. These controller instances compete to use a shared resource, known as the leader election lock. The instance of the controller that has access to the leader election lock resource at any given time is referred to as the leader.
+
+HA functionality is available on Knative for the `autoscaler-hpa`, `controller`, `activator` , `kourier-control`, `kourier-gateway`, `istio-pilot`, and `istio-gateway` components.
```
Comment (pmorie) [inline docs/install/config-ha.md:14]: I'm not sure if pilot and gateway are the right level of info to expose... Maybe 'and the controllers integrating Knative with Kourier and Istio ingresses' is better? I don't think many users will be understand exactly what these names mean. WDYT?

File: docs/install/config-ha.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+## Enabling leader election
+
+1. Enable leader election for the control plane controllers:
+```
+kubectl patch configmap/config-leader-election \
+  --namespace knative-serving \
+  --type merge \
+  --patch '{"data":{"enabledComponents": "controller,hpaautoscaler,certcontroller,istiocontroller,nscontroller"}}'
```
Comment (pmorie) [inline docs/install/config-ha.md:23]: This won't change Kourier, you would need to run a different command for that that touched the same configmap, but in the `kourier-system` namespace.

File: docs/install/config-ha.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+```
+
+1. Restart the controllers:
+
+Note: you will take a brief amount of control plane downtime during this step.
+
+```
+kubectl rollout restart deployment -n knative-serving
```
Comment (pmorie) [inline docs/install/config-ha.md:31]: same note here as above, this won't affect kourier if it is installed (however it will affect istio integration since that is deployed in the knative-serving ns and controlled via the serving LE config).

File: docs/install/config-ha.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+The following serving controller deployments can be scaled up once leader election is enabled:
+
+- `controller`
+- `autoscaler-hpa`
+- `networking-certmanager`
+- `networking-istio` (if Istio is installed)
+- `networking-ns-cert`
+
+```
+kubectl scale --replicas=2 <deployment-name>
```
Comment (pmorie) [inline docs/install/config-ha.md:54]: (in reference to earlier comment about exposing deployment names), I guess we do expose these here...

File: docs/install/config-ha.md (hunk 3 above)
Comment (vagababov) [reply docs/install/config-ha.md:14]: In general, as I proposed above, I'd just keep knative components.

Comment (abrennan89) [reply docs/install/config-ha.md:14]: Cool, will do. Thanks folks.

File: docs/install/config-ha.md (hunk 4 above)
Comment (abrennan89) [reply docs/install/config-ha.md:23]: Should we add this then? Could you provide the command and any other details please?

**Comment by abrennan89** [comment]
I'd keep the details about the Istio and Kourier components, and then maybe just have sub-sections for using Kourier, using Istio, and using none of them? @vagababov @pmorie wdyt?

File: docs/install/config-ha.md (hunk 3 above)
Comment (ZhiminXiang) [reply docs/install/config-ha.md:14]: `istio-pilot` and `istio-gateway` are not provided and controlled by Knative. They are provided by Istio. So better to exclude them.

If we want to include kourier-control, do we also want to include istio-controller and contour-controller?

File: docs/install/config-ha.md (hunk 4 above)
Comment (abrennan89) [reply docs/install/config-ha.md:23]: Just gonna remove all mentions of Kourier etc for now, we can do some follow up stuff to flesh out any docs we need for that later.

File: docs/config-ha.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
+---
+title: "Configuring high-availability components"
+linkTitle: "Configuring high-availability components"
+weight: 20
+type: "docs"
+---
+
+Active/passive high availability (HA) is a standard feature of Kubernetes APIs that helps to ensure that APIs stay operational if a disruption occurs. In an HA deployment, if an active controller crashes or is deleted, another controller is available to take over processing of the APIs that were being serviced by the controller that is now unavailable.
+
+Active/passive HA in Knative is available through leader election, which is enabled by default after the Knative Serving control plane is installed.
```
Comment (vagababov) [inline docs/config-ha.md:10]: Right now it isn't. Users need to enable it.
```suggestion
Active/passive HA in Knative is available through leader election, which is not enabled by default after the Knative Serving control plane is installed.
```

**Comment by vagababov** [comment]
Works for me.
/lgtm

**Comment by abrennan89** [comment]
@carieshmarie @mattmoor could someone review this for docs please? Thanks.

File: docs/config-ha.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
+---
+title: "Configuring high-availability components"
+linkTitle: "Configuring high-availability components"
+weight: 20
+type: "docs"
+---
+
+Active/passive high availability (HA) is a standard feature of Kubernetes APIs that helps to ensure that APIs stay operational if a disruption occurs. In an HA deployment, if an active controller crashes or is deleted, another controller is available to take over processing of the APIs that were being serviced by the controller that is now unavailable.
+
+Active/passive HA in Knative is available through leader election, which can be enabled after Knative Serving control plane is installed.
```
Comment (mattmoor) [inline docs/config-ha.md:10]: I was surprised these docs were pretty serving-centric given their location and the title.

e.g. How do I do the same for eventing?

File: docs/config-ha.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+When using a leader election HA pattern, instances of controllers are already scheduled and running inside the cluster before they are required. These controller instances compete to use a shared resource, known as the leader election lock. The instance of the controller that has access to the leader election lock resource at any given time is referred to as the leader.
+
+HA functionality is available on Knative for the `autoscaler-hpa`, `controller` and `activator` components.
```
Comment (mattmoor) [inline docs/config-ha.md:14]: This seems inconsistent with the list below?

Comment (mattmoor) [reply docs/config-ha.md:14]: Perhaps we should include a list of what's supported *and not supported* broken down by namespace?

As a user, I'd probably want to know what isn't HA (yet).

File: docs/config-ha.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+## Enabling leader election
```
Comment (mattmoor) [inline docs/config-ha.md:16]: I feel like we should clearly indicate a feature-state before asking folks to enable things.

Given that this is still opt-in and baking, and doesn't yet cover everything, I'd probably call this "alpha" still (cc @pmorie for thoughts, and since we haven't discussed a designation).

Personally, I'd love to more towards having this enabled everywhere by default and let these docs focus more on how to tune things (e.g. lease refreshes, etc).

File: docs/config-ha.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+  **NOTE:** You will experience temporary control plane downtime during this step.
+
+  When your controllers come back up, they should be running as leader-elected.
+  At this point, we've configured the controllers to use leader election and we
+  can scale the control plane up!
+
+1. After the controllers have been configured to use leader election, the control plane can be scaled up:
+```
+$ kubectl rollouts restart deployment -n knative-serving controller
```
Comment (mattmoor) [inline docs/config-ha.md:39]: I think you meant `scale` here?

Comment (mattmoor) [reply docs/config-ha.md:39]: Hmm, I see this below, but also a restart without the `controller` above.  Now I'm confused :)

One trick that we use in serving is: `--scale=0` -> `--scale=N` to restart things AND scale it up.  It shouldn't result in any more downtime, and could actually avoid fighting controllers as the old one shuts down and the new one comes up.

File: docs/config-ha.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+## Scaling the control plane
+
+The following serving controller deployments can be scaled up once leader election is enabled:
+
+- `controller`
+- `autoscaler-hpa`
+- `networking-certmanager`
+- `networking-istio` (if Istio is installed)
+- `networking-ns-cert`
```
Comment (mattmoor) [inline docs/config-ha.md:50]: This is also an optional install.

File: docs/config-ha.md [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+$ kubectl rollouts restart deployment -n knative-serving controller
+```
+
+## Scaling the control plane
+
+The following serving controller deployments can be scaled up once leader election is enabled:
+
+- `controller`
+- `autoscaler-hpa`
+- `networking-certmanager`
```
Comment (mattmoor) [inline docs/config-ha.md:48]: This is also an optional install.

File: docs/config-ha.md [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+```
+$ kubectl rollouts restart deployment -n knative-serving controller
+```
+
+## Scaling the control plane
+
+The following serving controller deployments can be scaled up once leader election is enabled:
+
+- `controller`
+- `autoscaler-hpa`
```
Comment (mattmoor) [inline docs/config-ha.md:47]: This is also an optional install

File: docs/config-ha.md [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+- `networking-certmanager`
+- `networking-istio` (if Istio is installed)
+- `networking-ns-cert`
+
+Scale up the deployment(s):
+```
+$ kubectl scale --replicas=2 <deployment-name>
```
Comment (mattmoor) [inline docs/config-ha.md:54]: Do we have guidance on choosing a replica count?

File: docs/config-ha.md (hunk 8 above)
Comment (lionelvillard) [reply docs/config-ha.md:10]: There is another page that is under serving but applies to eventing as well: https://knative.dev/docs/serving/accessing-metrics/

Should we move this under "Installing Knative" (which I think should be renamed to "Getting Started with Knative")?

Comment (mattmoor) [reply docs/config-ha.md:10]: I'd rather not move that more prominently as it builds on something we're turning down :)

Comment (abrennan89) [reply docs/config-ha.md:10]: Fair, I wasn't sure where to put it. I'll move it under Serving until we have something related to HA for Eventing.

File: docs/config-ha.md (hunk 9 above)
Comment (abrennan89) [reply docs/config-ha.md:14]: @pmorie can you provide details for this please?

File: docs/config-ha.md (hunk 10 above)
Comment (pmorie) [reply docs/config-ha.md:16]: I would be okay calling it 'alpha' in this PR, but we'll need to discuss a scheme for these things (hopefully we can just use alpha, beta, GA)

File: docs/config-ha.md (hunk 9 above)
Comment (abrennan89) [reply docs/config-ha.md:14]: I'm honestly not sure if including a list of components that are or aren't supported here makes sense if that list is likely to be in flux for now, docs maintenance wise. I'll add it, but let's revisit and discuss the future plans for HA docs soon.

File: docs/config-ha.md (hunk 15 above)
Comment (abrennan89) [reply docs/config-ha.md:54]: Added some extra info.

File: docs/config-ha.md [hunk 16]
```diff
@@ -0,0 +1,55 @@
+---
+title: "Configuring high-availability components"
+linkTitle: "Configuring high-availability components"
```
[the remark this answers is not in the document]
Comment (abrennan89) [reply docs/config-ha.md]: Yep, good to know re linktitle, I'll remove it :) awaiting some more confirmation about the last comment from Matt too.

**Comment by abrennan89** [comment]
/hold

Matt is checking some commands

**Comment by abrennan89** [comment]
> Is this still [WIP]?

@evankanderson should be good to go now
